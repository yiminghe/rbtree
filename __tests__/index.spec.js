import RBTree, { RBColor } from '../src/';

function format(n) {
  const key = n.key || '*';
  const val = n.val || '*';
  const content = key;
  return n.color === RBColor.RED ? content + ':0' : content + ':1';
}

function size(n) {
  return format(n).length;
}

describe('red-black-tree', () => {
  it('works', () => {
    const tree = new RBTree();

    let data = [27, 25, 22, 17, 15, 13, 11, 8, 6, 1];

    function check() {
      expect(tree.getBlackCount()).toMatchSnapshot();
      expect(tree.visualize(format, size)).toMatchSnapshot();
    }

    for (const d of data) {
      tree.insert(d, d);
      check();
    }

    // console.log(tree.visualize(format, size));

    for (const d of data) {
      expect(tree.find(d)).toBe(d);
    }

    expect(tree.lowerBound(9)).toBe(11);
    expect(tree.upperBound(18)).toBe(17);

    data = [11, 15, 22, 17, 6, 25, 8, 27, 1, 13];

    for (const d of data) {
      tree.delete(d);
      check();
    }
  });
});
