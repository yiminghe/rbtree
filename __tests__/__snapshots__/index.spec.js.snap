// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`red-black-tree works 1`] = `1`;

exports[`red-black-tree works 2`] = `
"    27:1
 *:1     *:1"
`;

exports[`red-black-tree works 3`] = `1`;

exports[`red-black-tree works 4`] = `
"            27:1
    25:0         *:1
 *:1     *:1"
`;

exports[`red-black-tree works 5`] = `1`;

exports[`red-black-tree works 6`] = `
"            25:1
    22:0            27:0
 *:1     *:1     *:1     *:1"
`;

exports[`red-black-tree works 7`] = `2`;

exports[`red-black-tree works 8`] = `
"                    25:1
            22:1            27:1
    17:0         *:1     *:1     *:1
 *:1     *:1"
`;

exports[`red-black-tree works 9`] = `2`;

exports[`red-black-tree works 10`] = `
"                            25:1
            17:1                    27:1
    15:0            22:0         *:1     *:1
 *:1     *:1     *:1     *:1"
`;

exports[`red-black-tree works 11`] = `2`;

exports[`red-black-tree works 12`] = `
"                                    25:1
                    17:0                    27:1
            15:1            22:1         *:1     *:1
    13:0         *:1     *:1     *:1
 *:1     *:1"
`;

exports[`red-black-tree works 13`] = `2`;

exports[`red-black-tree works 14`] = `
"                                            25:1
                            17:0                    27:1
            13:1                    22:1         *:1     *:1
    11:0            15:0         *:1     *:1
 *:1     *:1     *:1     *:1"
`;

exports[`red-black-tree works 15`] = `2`;

exports[`red-black-tree works 16`] = `
"                                    17:1
                    13:0                            25:0
            11:1            15:1            22:1            27:1
     8:0         *:1     *:1     *:1     *:1     *:1     *:1     *:1
 *:1     *:1"
`;

exports[`red-black-tree works 17`] = `2`;

exports[`red-black-tree works 18`] = `
"                                            17:1
                            13:0                            25:0
             8:1                    15:1            22:1            27:1
     6:0            11:0         *:1     *:1     *:1     *:1     *:1     *:1
 *:1     *:1     *:1     *:1"
`;

exports[`red-black-tree works 19`] = `3`;

exports[`red-black-tree works 20`] = `
"                                                    17:1
                                    13:1                            25:1
                     8:0                    15:1            22:1            27:1
             6:1            11:1         *:1     *:1     *:1     *:1     *:1     *:1
     1:0         *:1     *:1     *:1
 *:1     *:1"
`;

exports[`red-black-tree works 21`] = `3`;

exports[`red-black-tree works 22`] = `
"                                            17:1
                            13:1                            25:1
             6:0                    15:1            22:1            27:1
     1:1             8:1         *:1     *:1     *:1     *:1     *:1     *:1
 *:1     *:1     *:1     *:1"
`;

exports[`red-black-tree works 23`] = `3`;

exports[`red-black-tree works 24`] = `
"                                    17:1
             6:1                                    25:1
     1:1                    13:1            22:1            27:1
 *:1     *:1         8:0         *:1     *:1     *:1     *:1     *:1
                 *:1     *:1"
`;

exports[`red-black-tree works 25`] = `2`;

exports[`red-black-tree works 26`] = `
"                                    17:1
             6:0                            25:1
     1:1                    13:1         *:1        27:0
 *:1     *:1         8:0         *:1             *:1     *:1
                 *:1     *:1"
`;

exports[`red-black-tree works 27`] = `2`;

exports[`red-black-tree works 28`] = `
"                            13:1
             6:0                    25:1
     1:1             8:1         *:1        27:0
 *:1     *:1     *:1     *:1             *:1     *:1"
`;

exports[`red-black-tree works 29`] = `2`;

exports[`red-black-tree works 30`] = `
"                    13:1
     1:1                    25:1
 *:1         8:0         *:1        27:0
         *:1     *:1             *:1     *:1"
`;

exports[`red-black-tree works 31`] = `2`;

exports[`red-black-tree works 32`] = `
"                    13:1
     1:1                    27:1
 *:1         8:0         *:1     *:1
         *:1     *:1"
`;

exports[`red-black-tree works 33`] = `2`;

exports[`red-black-tree works 34`] = `
"            13:1
     1:1            27:1
 *:1     *:1     *:1     *:1"
`;

exports[`red-black-tree works 35`] = `1`;

exports[`red-black-tree works 36`] = `
"            13:1
     1:0         *:1
 *:1     *:1"
`;

exports[`red-black-tree works 37`] = `1`;

exports[`red-black-tree works 38`] = `
"    13:1
 *:1     *:1"
`;

exports[`red-black-tree works 39`] = `0`;

exports[`red-black-tree works 40`] = `"*:1"`;
